polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                None => panic!("unknown field {}", field),
+                Some(t) => t,
+            };
+
+            if let Type::Relationship {
+                other_class_tag,
+                my_field,
+                other_field,
+                ..
+            } = typ
+            {
+                constrain_var(result_set, types, vars, child, other_class_tag);
+
+                // If the constrained child var doesn't have any constraints on it, we don't need to
+                // constrain this var. Otherwise we're just saying field foo in all Foos which
+                // would fetch all Foos and not be good.
+                if let Some(child_result) = result_set.requests.remove(child) {
+                    if !child_result.constraints.is_empty() {
+                        result_set.requests.insert(child.to_owned(), child_result);
+                        request.constraints.push(Constraint {
+                            kind: ConstraintKind::In,
+                            field: my_field.clone(),
+                            value: ConstraintValue::Ref(Ref {
+                                field: Some(other_field.clone()),
+                                result_id: child.clone(),
+                            }),
+                        });
+                    } else {
+                        // Remove the id from the resolve_order too.
+                        result_set.resolve_order.pop();
+                    }
+                }
+
+                continue;
+            }
+            // Non relationship or unknown type info.
+            let mut contributed_constraints = false;
+            if let Some(value) = vars.eq_values.get(child) {
+                request.constraints.push(Constraint {
+                    kind: ConstraintKind::Eq,
+                    field: field.clone(),
+                    value: ConstraintValue::Term(value.clone()),
+                });
+                contributed_constraints = true;
+            }
+            if let Some(values) = vars.contained_values.get(child) {
+                for value in values {
+                    request.constraints.push(Constraint {
+                        kind: ConstraintKind::Contains,
+                        field: field.clone(),
+                        value: ConstraintValue::Term(value.clone()),
+                    });
+                }
+                contributed_constraints = true;
+            }
+            for eqf in vars
+                .field_relationships
+                .iter()
+                .filter(|r| r.0 == *parent && r.1 != *field && r.2 == *child)
+            {
+                request.constraints.push(Constraint {
+                    kind: ConstraintKind::Eq,
+                    field: field.clone(),
+                    value: ConstraintValue::Field(eqf.1.clone()),
+                });
+                contributed_constraints = true;
+            }
+            assert!(contributed_constraints);
```

Comment:
So we'd hit this assert if the relationship wasn't defined rn?

Reply:
Not sure, that's a good test to see what happens.